File: benchmarks/core-comparison/programs/validated-sum.c
#include <stdio.h>

static int validated_integer(double value) {
    const long long converted = (long long) value;
    return (double) converted == value;
}

static unsigned long long count_valid_integers(
    const volatile double *data,
    int length
) {
    unsigned long long accepted = 0;
    for (int index = 0; index < length; ++index) {
        accepted += (unsigned long long) validated_integer(data[index]);
    }
    return accepted;
}

int main(void) {
    volatile double values[] = { {{VALIDATION_VALUES}} };
    printf("%llu\n", count_valid_integers(values, {{COUNT}}));
    return 0;
}

File: benchmarks/core-comparison/programs/validated-sum.rs
#[derive(Clone, Copy)]
struct ValueError;

fn validated_integer(value: f64) -> Result<i64, ValueError> {
    if value.floor() == value {
        Ok(value as i64)
    } else {
        Err(ValueError)
    }
}

fn count_valid_integers(data: &[f64]) -> u64 {
    let mut accepted = 0_u64;
    for &value in data {
        match validated_integer(value) {
            Ok(_) => accepted += 1,
            Err(ValueError) => {}
        }
    }
    accepted
}

fn main() {
    let values = std::hint::black_box(vec![{{VALIDATION_VALUES}}]);
    println!("{}", count_valid_integers(&values));
}
